I can reproduce the Gradle test failures in my logs, but the cause seems unrelated to my changes:

> Caused by: org.gradle.internal.resource.transport.http.HttpErrorStatusCodeException: Could not HEAD 'https://repo.gradle.org/artifactory/jcenter/io/quarkus/quarkus-build-parent/999-SNAPSHOT/quarkus-build-parent-999-SNAPSHOT.pom'. Received status code 409 from server: Conflict

I'm investigating why in the world we'd be looking for this artifact on the JCenter mirror inside the Gradle Maven repository...

--- a/core/processor/src/main/java/io/quarkus/annotation/processor/ExtensionAnnotationProcessor.java
+++ b/core/processor/src/main/java/io/quarkus/annotation/processor/ExtensionAnnotationProcessor.java
@@ -25,6 +25,7 @@ import java.util.Collection;
 import java.util.Collections;
 import java.util.HashSet;
 import java.util.List;
+import java.util.Map;
 import java.util.Properties;
 import java.util.Set;
 import java.util.TreeSet;
@@ -37,6 +38,7 @@ import javax.annotation.processing.Filer;
 import javax.annotation.processing.RoundEnvironment;
 import javax.lang.model.SourceVersion;
 import javax.lang.model.element.AnnotationMirror;
+import javax.lang.model.element.AnnotationValue;
 import javax.lang.model.element.Element;
 import javax.lang.model.element.ElementKind;
 import javax.lang.model.element.ExecutableElement;
@@ -390,21 +392,21 @@ public class ExtensionAnnotationProcessor extends AbstractProcessor {
         for (Element e : clazz.getEnclosedElements()) {
             switch (e.getKind()) {
                 case FIELD: {
-                    if (isAnnotationPresent(e, Constants.ANNOTATION_CONFIG_ITEM, Constants.ANNOTATION_CONFIG_DOC_SECTION)) {
+                    if (isDocumentedConfigItem(e)) {
                         processFieldConfigItem((VariableElement) e, javadocProps, className);
                     }
                     break;
                 }
                 case CONSTRUCTOR: {
                     final ExecutableElement ex = (ExecutableElement) e;
-                    if (hasParameterAnnotated(ex, Constants.ANNOTATION_CONFIG_ITEM, Constants.ANNOTATION_CONFIG_DOC_SECTION)) {
+                    if (hasParameterDocumentedConfigItem(ex)) {
                         processCtorConfigItem(ex, javadocProps, className);
                     }
                     break;
                 }
                 case METHOD: {
                     final ExecutableElement ex = (ExecutableElement) e;
-                    if (hasParameterAnnotated(ex, Constants.ANNOTATION_CONFIG_ITEM, Constants.ANNOTATION_CONFIG_DOC_SECTION)) {
+                    if (hasParameterDocumentedConfigItem(ex)) {
                         processMethodConfigItem(ex, javadocProps, className);
                     }
                     break;
@@ -739,9 +741,40 @@ public class ExtensionAnnotationProcessor extends AbstractProcessor {
         return REMOVE_LEADING_SPACE.matcher(docComment).replaceAll("").trim();
     }
 
-    private static boolean hasParameterAnnotated(ExecutableElement ex, String... annotationNames) {
+    private static boolean isDocumentedConfigItem(Element element) {
+        boolean hasAnnotation = false;
+        for (AnnotationMirror annotationMirror : element.getAnnotationMirrors()) {
+            String annotationName = ((TypeElement) annotationMirror.getAnnotationType().asElement())
+                    .getQualifiedName().toString();
+            if (Constants.ANNOTATION_CONFIG_ITEM.equals(annotationName)) {
+                hasAnnotation = true;
+                Object generateDocumentation = getAnnotationAttribute(annotationMirror, "generateDocumentation()");
+                if (generateDocumentation != null && !(Boolean) generateDocumentation) {
+                    // Documentation is explicitly disabled
+                    return false;
+                }
+            } else if (Constants.ANNOTATION_CONFIG_DOC_SECTION.equals(annotationName)) {
+                hasAnnotation = true;
+            }
+        }
+        return hasAnnotation;
+    }
+
+    private static Object getAnnotationAttribute(AnnotationMirror annotationMirror, String attributeName) {
+        for (Map.Entry<? extends ExecutableElement, ? extends AnnotationValue> entry : annotationMirror
+                .getElementValues().entrySet()) {
+            final String key = entry.getKey().toString();
+            final Object value = entry.getValue().getValue();
+            if (attributeName.equals(key)) {
+                return value;
+            }
+        }
+        return null;
+    }
+
+    private static boolean hasParameterDocumentedConfigItem(ExecutableElement ex) {
         for (VariableElement param : ex.getParameters()) {
-            if (isAnnotationPresent(param, annotationNames)) {
+            if (isDocumentedConfigItem(param)) {
                 return true;
             }
         }

--- a/core/processor/src/main/java/io/quarkus/annotation/processor/generate_doc/ConfigDoItemFinder.java
+++ b/core/processor/src/main/java/io/quarkus/annotation/processor/generate_doc/ConfigDoItemFinder.java
@@ -166,6 +166,7 @@ class ConfigDoItemFinder {
             String hyphenatedFieldName = hyphenate(fieldName);
             String configDocMapKey = hyphenatedFieldName;
             boolean isDeprecated = false;
+            boolean generateDocumentation = true;
             ConfigDocSection configSection = new ConfigDocSection();
             configSection.setTopLevelGrouping(rootName);
             configSection.setWithinAMap(withinAMap);
@@ -182,12 +183,12 @@ class ConfigDoItemFinder {
                     for (Map.Entry<? extends ExecutableElement, ? extends AnnotationValue> entry : annotationMirror
                             .getElementValues().entrySet()) {
                         final String key = entry.getKey().toString();
-                        final String value = entry.getValue().getValue().toString();
+                        final Object value = entry.getValue().getValue();
                         if (annotationName.equals(ANNOTATION_CONFIG_DOC_MAP_KEY) && "value()".equals(key)) {
-                            configDocMapKey = value;
+                            configDocMapKey = value.toString();
                         } else if (annotationName.equals(ANNOTATION_CONFIG_ITEM)) {
                             if ("name()".equals(key)) {
-                                switch (value) {
+                                switch (value.toString()) {
                                     case HYPHENATED_ELEMENT_NAME:
                                         name = parentName + DOT + hyphenatedFieldName;
                                         break;
@@ -198,9 +199,11 @@ class ConfigDoItemFinder {
                                         name = parentName + DOT + value;
                                 }
                             } else if ("defaultValue()".equals(key)) {
-                                defaultValue = value;
+                                defaultValue = value.toString();
                             } else if ("defaultValueDocumentation()".equals(key)) {
-                                defaultValueDoc = value;
+                                defaultValueDoc = value.toString();
+                            } else if ("generateDocumentation()".equals(key)) {
+                                generateDocumentation = (Boolean) value;
                             }
                         }
                     }
@@ -218,6 +221,9 @@ class ConfigDoItemFinder {
 
             if (isDeprecated) {
                 continue; // do not include deprecated config items
+            }
+            if (!generateDocumentation) {
+                continue; // documentation for this item was explicitly disabled
             }
 
             if (name == null) {

--- a/core/runtime/src/main/java/io/quarkus/runtime/annotations/ConfigItem.java
+++ b/core/runtime/src/main/java/io/quarkus/runtime/annotations/ConfigItem.java
@@ -61,4 +61,15 @@ public @interface ConfigItem {
      * @return the default value documentation
      */
     String defaultValueDocumentation() default "";
+
+    /**
+     * Specify whether documentation should be generated for this config item.
+     * <p>
+     * This is only useful in very niche use cases where the build-time and runtime config
+     * both contain a configuration property with the exact same name,
+     * and we only want to generate documentation from one of them.
+     *
+     * @return whether documentation should be generated
+     */
+    boolean generateDocumentation() default true;
 }

